blockchain/chain.go

```diff
@@ -1679,6 +1674,156 @@ func (b *BlockChain) HeaderByHash(hash *chainhash.Hash) (wire.BlockHeader, error
+	}
+
+	return node.Header(), nil
+}
+
+// BlockByHash searches the internal chain block stores and the database in an
+// attempt to find the requested block and returns it.  This function returns
+// blocks regardless of whether or not they are part of the main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) BlockByHash(hash *chainhash.Hash) (*dcrutil.Block, error) {
+	node := b.index.LookupNode(hash)
+	if node == nil {
+		return nil, fmt.Errorf("block %s is not known", hash)
+	}
+
+	// Return the block from either cache or the database.
+	return b.fetchBlockByNode(node)
+}
+
+// BlockByHeight returns the block at the given height in the main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) BlockByHeight(height int64) (*dcrutil.Block, error) {
+	b.heightLock.RLock()
+	node := b.mainNodesByHeight[height]
+	b.heightLock.RUnlock()
+	if node == nil {
+		str := fmt.Sprintf("no block at height %d exists", height)
+		return nil, errNotInMainChain(str)
+	}
+
+	// Return the block from either cache or the database.  Note that this is
+	// not using fetchMainChainBlockByNode since the main chain check has
+	// already been done.
+	return b.fetchBlockByNode(node)
+}
+
+// MainChainHasBlock returns whether or not the block with the given hash is in
+// the main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) MainChainHasBlock(hash *chainhash.Hash) bool {
+	node := b.index.LookupNode(hash)
+	b.chainLock.RLock()
+	hasBlock := node != nil && node.inMainChain
+	b.chainLock.RUnlock()
+	return hasBlock
+}
+
+// BlockHeightByHash returns the height of the block with the given hash in the
+// main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) BlockHeightByHash(hash *chainhash.Hash) (int64, error) {
+	node := b.index.LookupNode(hash)
+	b.chainLock.RLock()
+	if node == nil || !node.inMainChain {
+		b.chainLock.RUnlock()
+		str := fmt.Sprintf("block %s is not in the main chain", hash)
+		return 0, errNotInMainChain(str)
+	}
+	b.chainLock.RUnlock()
+
+	return node.height, nil
+}
+
+// BlockHashByHeight returns the hash of the block at the given height in the
+// main chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) BlockHashByHeight(height int64) (*chainhash.Hash, error) {
+	b.heightLock.RLock()
+	node := b.mainNodesByHeight[height]
+	b.heightLock.RUnlock()
+	if node == nil {
+		str := fmt.Sprintf("no block at height %d exists", height)
+		return nil, errNotInMainChain(str)
+	}
+
+	return &node.hash, nil
+}
+
+// HeightRange returns a range of block hashes for the given start and end
+// heights.  It is inclusive of the start height and exclusive of the end
+// height.  The end height will be limited to the current main chain height.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) HeightRange(startHeight, endHeight int64) ([]chainhash.Hash, error) {
```

Comment:
perhaps document that this is the half open range [startHeight, endHeight)

Reply:
It says `It is inclusive of the start height and exclusive of the end height` already.  I figured that would make it clear.  I'll go ahead and expand on it a bit to include the blurb you suggest.

Reply:
Updated.